File: packages/engine/src/lib/__tests__/determinism.test.ts
import { describe, expect, it } from 'vitest';
import { GameActions } from '../game-actions';
import { GameEngine } from '../game-engine';
import {
  createTestCard,
  createTestPlayer,
  createTestState,
  markPlayersReady,
  toPile,
  unsafeReduce,
} from './test-helpers';

/**
 * Replay and cross-implementation parity both rest on one property: the same state and
 * the same action always produce the same next state, ids and rngState included.
 */
describe('engine determinism', () => {
  it('produces identical results for two reductions of equal inputs', () => {
    const state = createTestState({
      phase: 'playing',
      subPhase: 'idle',
      drawPile: toPile([
        createTestCard('5', 'd1'),
        createTestCard('9', 'd2'),
        createTestCard('K', 'd3'),
      ]),
      rngState: 12345,
    });

    const action = GameActions.drawCard('p1');
    const first = GameEngine.reduce(state, action);
    const second = GameEngine.reduce(state, action);

    expect(first.success).toBe(true);
    expect(second.success).toBe(true);
    expect(JSON.stringify(second.state)).toBe(JSON.stringify(first.state));
    expect(second.state.rngState).toBe(first.state.rngState);
  });

  it('leaves rngState untouched when an action consumes no randomness', () => {
    const state = createTestState({
      phase: 'playing',
      subPhase: 'idle',
      drawPile: toPile([createTestCard('5', 'd1'), createTestCard('9', 'd2')]),
      rngState: 999,
    });

    const result = GameEngine.reduce(state, GameActions.drawCard('p1'));

    expect(result.success).toBe(true);
    expect(result.state.rngState).toBe(999);
  });

  it('advances rngState reproducibly when the draw pile is reshuffled', () => {
    // The reshuffle fires from advanceTurnAfterTossIn once the draw pile is down to one
    // card, which is the engine's only consumer of randomness. Driven through real turns,
    // mirroring Scenario 07 in scenarios.test.ts.
    const buildState = (rngState: number) =>
      createTestState({
        subPhase: 'idle',
        currentPlayerIndex: 0,
        turnNumber: 1,
        players: [
          createTestPlayer('p1', 'Player 1', true, [
            createTestCard('2', 'p1c1'),
          ]),
          createTestPlayer('p2', 'Player 2', false, [
            createTestCard('3', 'p2c1'),
          ]),
          createTestPlayer('p3', 'Player 3', false, [
            createTestCard('4', 'p3c1'),
          ]),
          createTestPlayer('p4', 'Player 4', false, [
            createTestCard('5', 'p4c1'),
          ]),
        ],
        drawPile: toPile([
          createTestCard('2', 'draw1'),
          createTestCard('3', 'draw2'),
          createTestCard('4', 'draw3'),
          createTestCard('5', 'draw4'),
        ]),
        discardPile: toPile([
          createTestCard('6', 'discard1'),
          createTestCard('6', 'discard2'),
          createTestCard('6', 'discard3'),
        ]),
        rngState,
      });

    const playToReshuffle = (rngState: number) => {
      let state = buildState(rngState);
      for (const playerId of ['p1', 'p2', 'p3']) {
        state = unsafeReduce(state, GameActions.drawCard(playerId));
        state = unsafeReduce(state, GameActions.discardCard(playerId));
        state = markPlayersReady(state, ['p1', 'p2', 'p3', 'p4']);
      }
      return state;
    };

    const runA = playToReshuffle(42);
    const runB = playToReshuffle(42);
    const different = playToReshuffle(4242);

    // The reshuffle happened.
    expect(runA.drawPile.length).toBe(6);
    expect(runA.discardPile.length).toBe(1);

    // Same seed in, same shuffled order out.
    expect(runB.drawPile.toArray()).toEqual(runA.drawPile.toArray());
    expect(runB.rngState).toBe(runA.rngState);

    // The generator actually moved, and a different seed gives a different order.
    expect(runA.rngState).not.toBe(42);
    expect(different.drawPile.toArray()).not.toEqual(runA.drawPile.toArray());
  });

  it('mints toss-in queued card ids without a clock', () => {
    const state = createTestState({
      phase: 'playing',
      subPhase: 'toss_queue_active',
      turnNumber: 7,
      activeTossIn: {
        ranks: ['9'],
        initiatorId: 'p2',
        originalPlayerIndex: 0,
        participants: ['p2'],
        queuedActions: [{ playerId: 'p2', rank: '9', position: 0 }],
        waitingForInput: false,
        playersReadyForNextTurn: ['p1', 'p3', 'p4'],
      },
    });

    const first = unsafeReduce(state, GameActions.playerTossInFinished('p2'));
    const second = unsafeReduce(state, GameActions.playerTossInFinished('p2'));

    expect(first.pendingAction?.card.id).toBe(second.pendingAction?.card.id);
    expect(first.pendingAction?.card.id).not.toMatch(/\d{13}/); // no epoch millis
    expect(first.pendingAction?.card.id).toContain('tossin_queued_7_p2_9');
  });
});

File: shared/engine/src/commonTest/kotlin/game/vinto/engine/DeterminismTest.kt
package game.vinto.engine

import game.vinto.shapes.GameAction
import game.vinto.shapes.GamePhase
import game.vinto.shapes.GameState
import game.vinto.shapes.GameSubPhase
import game.vinto.shapes.PlayerIdPayload
import game.vinto.shapes.Rank
import game.vinto.shapes.TossInAction
import game.vinto.shapes.canonicalizeGameState
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

/**
 * Determinism, ported from `legacy-web/packages/engine/src/lib/__tests__/determinism.test.ts`.
 *
 * Replay and cross-language parity both rest on one property: the same state and the same
 * action always produce the same next state — ids and `rngState` included. Everything else
 * in the migration is downstream of this holding.
 */
class DeterminismTest {

    @Test
    fun twoReductionsOfEqualInputsAgreeExactly() {
        val state = testState(
            drawPile = pileOf(
                testCard(Rank.FIVE, "d1"),
                testCard(Rank.NINE, "d2"),
                testCard(Rank.KING, "d3"),
            ),
            rngState = 12345,
        )

        val first = GameEngine.reduce(state, drawCard("p1"))
        val second = GameEngine.reduce(state, drawCard("p1"))

        assertTrue(first is ReduceResult.Success)
        assertTrue(second is ReduceResult.Success)
        assertEquals(canonicalizeGameState(first.state), canonicalizeGameState(second.state))
        assertEquals(first.state.rngState, second.state.rngState)
    }

    @Test
    fun anActionThatNeedsNoRandomnessLeavesTheGeneratorAlone() {
        val state = testState(
            drawPile = pileOf(testCard(Rank.FIVE, "d1"), testCard(Rank.NINE, "d2")),
            rngState = 999,
        )

        val result = GameEngine.reduce(state, drawCard("p1"))

        assertTrue(result is ReduceResult.Success)
        assertEquals(999L, result.state.rngState)
    }

    @Test
    fun aReshuffleAdvancesTheGeneratorReproducibly() {
        // The reshuffle fires from advanceTurnAfterTossIn once the draw pile is down to one
        // card, and it is the engine's only consumer of randomness. Driven through real
        // turns rather than called directly, so the trigger is under test too.
        fun buildState(rngState: Long) = testState(
            subPhase = GameSubPhase.IDLE,
            turnNumber = 1,
            players = listOf(
                testPlayer("p1", "Player 1", isHuman = true, cards = listOf(testCard(Rank.TWO, "p1c1"))),
                testPlayer("p2", "Player 2", isHuman = false, cards = listOf(testCard(Rank.THREE, "p2c1"))),
                testPlayer("p3", "Player 3", isHuman = false, cards = listOf(testCard(Rank.FOUR, "p3c1"))),
                testPlayer("p4", "Player 4", isHuman = false, cards = listOf(testCard(Rank.FIVE, "p4c1"))),
            ),
            drawPile = pileOf(
                testCard(Rank.TWO, "draw1"), testCard(Rank.THREE, "draw2"),
                testCard(Rank.FOUR, "draw3"), testCard(Rank.FIVE, "draw4"),
            ),
            discardPile = pileOf(
                testCard(Rank.SIX, "discard1"),
                testCard(Rank.SIX, "discard2"),
                testCard(Rank.SIX, "discard3"),
            ),
            rngState = rngState,
        )

        fun playToReshuffle(rngState: Long): GameState {
            var state = buildState(rngState)
            for (playerId in listOf("p1", "p2", "p3")) {
                state = unsafeReduce(state, drawCard(playerId))
                state = unsafeReduce(state, discardCard(playerId))
                state = markPlayersReady(state, listOf("p1", "p2", "p3", "p4"))
            }
            return state
        }

        val runA = playToReshuffle(42)
        val runB = playToReshuffle(42)
        val different = playToReshuffle(4242)

        assertEquals(6, runA.drawPile.size, "the reshuffle did not happen")
        assertEquals(1, runA.discardPile.size)

        assertEquals(runA.drawPile.cards.map { it.id }, runB.drawPile.cards.map { it.id })
        assertEquals(runA.rngState, runB.rngState)

        assertNotEquals(42L, runA.rngState, "the generator never moved")
        assertNotEquals(
            runA.drawPile.cards.map { it.id },
            different.drawPile.cards.map { it.id },
            "a different seed produced the same order",
        )
    }

    @Test
    fun queuedTossInCardIdsAreMintedWithoutAClock() {
        val state = testState(
            phase = GamePhase.PLAYING,
            subPhase = GameSubPhase.TOSS_QUEUE_ACTIVE,
            turnNumber = 7,
            activeTossIn = tossIn(
                ranks = listOf(Rank.NINE),
                initiatorId = "p2",
                participants = listOf("p2"),
                queuedActions = listOf(TossInAction(playerId = "p2", rank = Rank.NINE, position = 0)),
                playersReadyForNextTurn = listOf("p1", "p3", "p4"),
            ),
        )

        val finished = GameAction.PlayerTossInFinished(PlayerIdPayload("p2"))
        val first = unsafeReduce(state, finished)
        val second = unsafeReduce(state, finished)

        val id = first.pendingAction?.card?.id
        assertEquals(id, second.pendingAction?.card?.id)
        assertTrue(id?.contains("tossin_queued_7_p2_9") == true, "unexpected id: $id")
        assertTrue(
            id?.let { Regex("\\d{13}").containsMatchIn(it) } == false,
            "the id carries what looks like epoch millis: $id",
        )
    }
}
